module/applications/actor/api/base-actor-sheet.mjs

```diff
@@ -542,78 +568,47 @@ export default class BaseActorSheet extends PrimarySheetMixin(
+      if ( !levels.length ) continue;
+      if ( !spellbook.spell0 && prog.cantrips ) registerSection("spell0", 0, CONFIG.DND5E.spellcasting.spell);
+      if ( prog.separate ) {
+        for ( let lvl = 1; lvl <= Math.max(...levels); lvl++ ) {
+          registerSection(prog.spellSlotKey(lvl), lvl, prog);
+        }
+      } else {
+        const [lvl] = levels;
+        registerSection(prog.spellSlotKey(lvl), lvl, prog);
       }
     }
 
-    // Create spellbook sections for all alternative spell preparation modes that have spell slots.
-    for ( const [k, v] of Object.entries(CONFIG.DND5E.spellPreparationModes) ) {
-      if ( !(k in levels) || !v.upcast || !levels[k].max ) continue;
-
-      if ( !spellbook["0"] && v.cantrips ) registerSection("spell0", 0, CONFIG.DND5E.spellLevels[0]);
-      const l = levels[k];
-      const level = game.i18n.localize(`DND5E.SpellLevel${l.level}`);
-      const label = `${v.label} — ${level}`;
-      registerSection(k, sections[k], label, {
-        prepMode: k,
-        value: l.value,
-        max: l.max,
-        override: l.override,
-        config: v
-      });
-    }
-
     // Iterate over every spell item, adding spells to the spellbook by section
     (context.itemCategories.spells ?? []).forEach(spell => {
-      const mode = spell.system.preparation.mode || "prepared";
-      let s = spell.system.level || 0;
-      const sl = `spell${s}`;
+      let mode = spell.system.preparation.mode;
+      if ( !(mode in CONFIG.DND5E.spellcasting) ) mode = "atwill";
+
+      const config = CONFIG.DND5E.spellcasting[mode];
+      const spellLevel = spell.system.level || 0;
+      let sectionKey = config.spellSlotKey(spellLevel);
 
       // Spells from items
       if ( spell.getFlag("dnd5e", "cachedFor") ) {
-        s = "item";
+        sectionKey = "item";
         if ( !spell.system.linkedActivity?.displayInSpellbook ) return;
-        if ( !spellbook[s] ) {
-          registerSection(null, s, game.i18n.localize("DND5E.CAST.SECTIONS.Spellbook"));
-          spellbook[s].order = 1000;
-        }
-      }
-
-      // Specialized spellcasting modes (if they exist)
-      else if ( mode in sections ) {
-        s = sections[mode];
-        if ( !spellbook[s] ) {
-          const l = levels[mode] || {};
-          const config = CONFIG.DND5E.spellPreparationModes[mode];
-          registerSection(mode, s, config.label, {
-            prepMode: mode,
-            value: l.value,
-            max: l.max,
-            override: l.override,
-            config: config
-          });
+        if ( !spellbook[sectionKey] ) {
+          registerSection(sectionKey);
+          spellbook[sectionKey].order = 10000;
         }
       }
 
       // Sections for higher-level spells which the caster "should not" have, but spell items exist for
-      else if ( !spellbook[s] ) {
-        registerSection(sl, s, CONFIG.DND5E.spellLevels[s], {levels: levels[sl]});
+      else if ( !spellbook[sectionKey] ) {
```

Comment:
[... ] otherwise they're tossed into a section corresponding to a key in `actor.system.spells` or one of the "static" spellcastings. Special handling for `"spell0"` which is where all cantrips go, except for static ones (as is current).